File: fun_avar.m
function [avar, taus] = fun_avar(data, m, fs)
% [avar, taus] = fun_avar(data, m, fs) returns the Allan
% variance and the correlation time array taus.
%
% when compared to the MATLAB allanvar function there is a difference in
% running time, this is 6.5 longer and the error is negligible 
% 
% Author: stnk
    
    % Correlation time array
    taus = m/fs;
    % Half vectorized approach of equation from
    % IEEE 952-2020, p. 75
    N = length(data);
    data_sum = cumsum(data)/fs;
    avar = zeros(length(m), 1);
    for k = 1:length(m)
        avar(k) = sum((data_sum(k+2*m(k):N) ...
                        - 2*data_sum(k+m(k):N-m(k)) ...
                        + data_sum(k:N-2*m(k))).^2 );
    end
    avar = avar./(2.*taus.^2 .* (N-2*m));

end

File: julia/fun_avar.jl
function fun_avar(Ω, fs, m)
# Homemade function to calculate the Allan variance.
# It is 30 times longer than the function in AllanDeviations 

    if length(Ω)<3
        error("Allan variance needs at least 3 data points")
    end
    
    θ = cumsum(Ω)/fs;
    N = length(Ω)

    tau0 = 1/fs;
    taus = m*tau0;      # Cluster durations

    avar = zeros(length(m),1);
    # Homemade Allan variance calculation
    for (k, τ) in enumerate(m)
        avar[k] = sum( (θ[k+2*τ:N]
                        - 2*θ[k+τ:N-τ]
                        + θ[k:N-2*τ]).^2);
        avar[k] = avar[k]/(2* (τ/fs)^2 * (N-2*τ));
    end
        
    (taus, avar)
end